### ISSUE TITLE
Statically-linked builds crash with segmentation fault on `export a-b=c-d`

### ISSUE THREAD

**Issue by wtdcode** [fish-shell/fish-shell#5289]
Fish version: 2.7.1

System: Linux kali 4.16.0-kali2-amd64 #1 SMP Debian 4.16.12-1kali1 (2018-05-28) x86_64 GNU/Linux

How to reproduce:

```shell
~ # sh -c 'env HOME=$(mktemp -d) fish'
Welcome to fish, the friendly interactive shell
root@kali /root# export a-b=c-d
Segmentation fault
~ #
```

But bash will tell me `a-b` is not a valid identifier.

**Comment by wtdcode** [comment]
@faho 
It is `LANG=zh_CN.UTF-8`. BTW, on my mac, it outputs
```
~ $ export a-b=c-d
set: Variable name 'a-b' is not valid. See `help identifiers`.
/usr/local/Cellar/fish/2.7.1/share/fish/functions/export.fish (line 18):
                    set -gx $v[1] $v[2]
                    ^
in function 'export'
	called on standard input
	with parameter list 'a-b=c-d'


       setset - display and change shell variables.
        -

   Synopsis
       set [SCOPE_OPTIONS]
       set [OPTIONS] VARIABLE_NAME VALUES...
       set [OPTIONS] VARIABLE_NAME[INDICES]... VALUES...
       set ( -q | --query ) [SCOPE_OPTIONS] VARIABLE_NAMES...
       set ( -e | --erase ) [SCOPE_OPTIONS] VARIABLE_NAME
       set ( -e | --erase ) [SCOPE_OPTIONS] VARIABLE_NAME[INDICES]...
       set ( -S | --show ) [SCOPE_OPTIONS] [VARIABLE_NAME]...
```

and the locale is also `zh_CN.UTF-8`

**Comment by mqudsi** [comment]
Yeah, the original bug also had spotty repro depending on locale. 

@faho, in #4974 you wrote

> Both of these translations are marked as "fuzzy" - we no longer use those. See e.g. #4847.
> 
> (I have no idea why this even attempts to use translations with different formats)

But for `zh_CN`, the translation (on line 1900) is not marked as fuzzy (but it's also empty, so I don't know how that ends up working).

**Comment by faho** [comment]
>But for zh_CN, the translation (on line 1900) is not marked as fuzzy (but it's also empty, so I don't know how that ends up working).

@mqudsi: In 2.7.1, it was:

```po
#: src/builtin.h:61
#, fuzzy, c-format
msgid "%ls: Variable name '%ls' is not valid. See `help identifiers`.\n"
msgstr "%ls：种子值 “%ls” 不是一个有效的数字\n"
```

(on line 1777)

Unfortunately the problem there was that the specifiers didn't match, and that doesn't appear to be the case here. I can't get this to reproduce (in 2.7.1 or master), but I can't even get that translation to trigger. It just always seems to print the english message.

**Comment by wtdcode** [comment]
@faho @mqudsi
I have built the fish from 2.7.1 release and master, both crashed with
```
/t/t/bin # export a-b=c-d                                                          fish: 'fish' terminated by signal SIGSEGV (Address boundary error)
```
But when I try to gdb it, it exit normally. Any idea about how to debug it by myself?

**Comment by mqudsi** [comment]
@wtdcode start fish in another terminal and start gdb in another terminal instructing it to attach to the already-started fish instance via `gdp -p fish_pid` substituting fish’s pid (`echo $fish_pid` will give you that under `master`) then try reproducing back in the fish window.

**Comment by wtdcode** [comment]
@mqudsi @faho 
The exception was thrown from [common.h](https://github.com/fish-shell/fish-shell/blob/fd13043340d714f70cb5d0d753c077d9bdb0f5e8/src/common.h#L553).
It seems that something is wrong with multithreading but I am not so familiar with that. Here I take two screenshots.
<img width="1595" alt="snipaste_2018-10-29_17-01-37" src="https://user-images.githubusercontent.com/30623163/47640168-003b4480-db9e-11e8-854a-a98323fb20cb.png">
<img width="1605" alt="snipaste_2018-10-29_17-03-51" src="https://user-images.githubusercontent.com/30623163/47640216-1cd77c80-db9e-11e8-948e-ce3ab50af317.png">

**Comment by faho** [comment]
>It seems that something is wrong with multithreading

Not quite. Fish uses multithreading, which means a running fish is made up of multiple threads, which you then see here.

However, the error seems to be simply one thread doing something naughty which triggers a segfault.

What would help us would be the entire call stack for all the threads. That should tell us where that one thread goes wrong.

Ideally, we'd be able to reproduce the crash, meaning that we could trigger it at will, but we're missing some information there - I thought it was down to language, but that doesn't seem to be the case. So the call stack might lead us somewhere.

Alternatively, you can try fish from current git master - there have been a whole bunch of changes, so if it doesn't crash it's most likely fixed.

**Comment by wtdcode** [comment]
@faho 
The version I built and debugged is from the source of master branch, not 2.7.1 tarball. Moreover, it now crashed when I input `export a`.
Sorry for forgetting to post full stacktrace:
```
#0  0x0000000000000000 in ?? ()
#1  0x000000000059ffdc in std::condition_variable::wait(std::unique_lock<std::mutex>&) ()
#2  0x00000000004a7d27 in iothread_perform_on_main(std::function<void ()>&&) (func=...) at src/iothread.cpp:339
#3  0x0000000000451654 in completer_t::<lambda(std::function<void()>&&)>::operator() (func=..., __closure=<synthetic pointer>) at src/complete.cpp:883
#4  completer_t::complete_param (this=0x7f81eace6940, cmd_orig=..., popt=..., str=..., use_switches=true) at src/complete.cpp:899
#5  0x00000000004522fd in completer_t::<lambda(const wcstring&, const wcstring&, size_t)>::operator() (depth=<optimized out>, cmdline=..., cmd=..., __closure=0x7f81e4001040) at src/complete.cpp:1555
#6  std::_Function_handler<void(const std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&, const std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&, long unsigned int), completer_t::perform()::<lambda(const wcstring&, const wcstring&, size_t)> >::_M_invoke(const std::_Any_data &, const std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > &, const std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > &, unsigned long &&) (__functor=..., __args#0=..., __args#1=..., __args#2=<optimized out>) at /usr/include/c++/7/bits/std_function.h:316
#7  0x00000000004530a4 in std::function<void (std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, unsigned long)>::operator()(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, unsigned long) const (__args#2=<optimized out>, __args#1=..., __args#0=..., this=0x7f81eace6890) at /usr/include/c++/7/bits/std_function.h:706
#8  walk_wrap_chain (command_line=..., command_range=..., command_range@entry=..., visitor=..., depth=depth@entry=0) at src/complete.cpp:1307
#9  0x0000000000456a9a in completer_t::perform (this=this@entry=0x7f81eace6940) at src/complete.cpp:1560
#10 0x0000000000456f8c in complete (cmd_with_subcmds=..., out_comps=out_comps@entry=0x7f81eace6a00, flags=flags@entry=1) at src/complete.cpp:1591
#11 0x00000000004d5bfd in <lambda()>::operator()(void) const (__closure=0x28aaae0) at src/reader.cpp:1332
#12 0x00000000004d5d1c in std::_Function_handler<autosuggestion_result_t(), get_autosuggestion_performer(const wcstring&, size_t, history_t*)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/7/bits/std_function.h:302
#13 0x00000000004dcee1 in std::function<autosuggestion_result_t ()>::operator()() const (this=0x281aef8) at /usr/include/c++/7/bits/std_function.h:706
#14 int _iothread_trampoline<autosuggestion_result_t>::perform<std::function<autosuggestion_result_t ()>, void (*)(autosuggestion_result_t)>(std::function<autosuggestion_result_t ()> const&, void (* const&)(autosuggestion_result_t))::{lambda()#1}::operator()() const (__closure=0x281aef0) at src/iothread.h:44
#15 std::_Function_handler<void (), int _iothread_trampoline<autosuggestion_result_t>::perform<std::function<autosuggestion_result_t ()>, void (*)(autosuggestion_result_t)>(std::function<autosuggestion_result_t ()> const&, void (* const&)(autosuggestion_result_t))::{lambda()#1}>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/include/c++/7/bits/std_function.h:316
#16 0x00000000004a6ffa in std::function<void ()>::operator()() const (this=0x7f81eace6c80) at /usr/include/c++/7/bits/std_function.h:706
#17 iothread_worker (unused=<optimized out>) at src/iothread.cpp:123
#18 0x00000000004fdbba in start_thread (arg=0x7f81eace7700) at pthread_create.c:463
#19 0x000000000068082f in clone ()
```
As for how to reproduce it, here is how my g++ was configured.

```
~/f/f/bin # g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 7.3.0-25' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Debian 7.3.0-25) 
```

**Comment by faho** [comment]
> Sorry for forgetting to post full stacktrace:

That's the wrong thread again. That looks like the autosuggestion thread, which tells the main thread to do some completions.

--------------

Also, can you reproduce this with `set -gx` in place of `export`? Does `set -gx a-b c-d` crash?

**Comment by mqudsi** [comment]
@wtdcode `thread apply all bt` will print the stack trace of all threads.

**Comment by wtdcode** [comment]
@faho @mqudsi 
> Moreover, it now crashes when I input `export a`.
This was happening when I took the two screenshots. In other word, it does crash when fish is trying to give me the autosuggestion `export a-b=c-d` when I just type `export a`, so I post the stacktrace of the autosuggestion thread.
In fact, in another machine running kali with g++8.2.0 and `LANG=en_US.UTF-8`, things become interesting.
```
/t/f/bin # ./fish
Welcome to fish, the friendly interactive shell
/t/f/bin # export a-b=c-d
set: Variable name 'a-b' is not valid. See `help identifiers`.
/tmp/fishbuild/share/fish/functions/export.fish (line 19): 
                    set -gx $v[1] $v[2]
                    ^
in function 'export'
        called on standard input
        with parameter list 'a-b=c-d'

/t/f/bin # g+= fish: “./fish” terminated by signal SIGSEGV (Address boundary error)
/t/f/bin # 
```
and
```
/t/f/bin # ./fish
Welcome to fish, the friendly interactive shell
/t/f/bin # export a-vfish: “./fish” terminated by signal SIGSEGV (Address boundary error)
```
It seems that fish is likely to crash when it is trying to give the autosuggestions. Then I try to  build it dynamically linked, everything works well.
So I find the problem should be '**fish is likely to crash when giving autosuggestions if statically linked**'.

**Comment by faho** [comment]
@wtdcode: How exactly are you building fish?

I can't even reproduce this on Kali, so clearly I must be missing some steps.

**Comment by wtdcode** [comment]
@faho 
```shell
env LDFLAGS=-static ./configure --prefix=/some/where/else
make -j8
make install
```
That's all.

**Comment by zanchey** [comment]
I've not tried recently, but statically-linked builds never worked when I tried them in the past.

**Comment by wtdcode** [comment]
@zanchey 
In fact, I have done some work on crosscompiling the fish to OpenWrt ( and works well), so when it crashed at first time on my kali machine I compiled it statically-linked habitually.
Even though the issue can be resolved by dynamic linking, I am a bit curious about why things go wrong statically-linked.

**Comment by mqudsi** [comment]
I don't think it's a problem with static linking, I imagine it's a memory access violation that static linking exposes.

**Comment by zanchey** [comment]
Yes, that's probably the case.

**Comment by wtdcode** [comment]
So anyone reproduces it successfully?

**Comment by faho** [comment]
Yes, I have been able to get it to crash when statically linking on a debian container. The stack trace looks just like yours, which is weird since it seems to crash in  `std::`.

Is this something like https://softwareengineering.stackexchange.com/questions/262195/whats-wrong-with-statically-linking-the-stl-into-multiple-shared-libraries?

**Comment by wtdcode** [comment]
@faho 
I have read that post, so do you mean the other thread is waiting an non-exist lock(which should exist in another copy of STL)?

**Comment by ridiculousfish** [comment]
@faho is it easy to provide simple instructions to repro with your container? I'm totally new to Docker.

**Comment by faho** [comment]
>is it easy to provide simple instructions to repro with your container?

I used systemd-nspawn, not docker, so the instructions are [on the arch wiki](https://wiki.archlinux.org/index.php/Systemd-nspawn#Create_a_Debian_or_Ubuntu_environment). I picked Debian Stable. Then `apt install git autoconf make libncurses-dev` and so on, git clone fish, and build [as above](https://github.com/fish-shell/fish-shell/issues/5289#issuecomment-434629794).

I'm not actually sure a debian environment is necessary, I just needed a container or something because arch (for some reason) strips the static libraries and so I'd have to rebuild ncurses myself.

Really, my guess is that just linking statically makes this blow up.
